Add tests for CLI utility.

ezekiel opened the pull request:
If CLI output format changes, must re-generate golden files.

This change is somewhat large and complex. Certainly I would value any sort of feedback/discussion about these new tests.
This change increases the time required for tests to run.

ezekiel commented:
Maybe it would make more sense to store the "golden" expected-output files in a dot directory, like `testdata/.golden/` ?

vanbroup commented:
What about using a filename such as `1024cert.pem.test` or `1024cert.pem.expected` instead of a hidden directory? This would keep the tested file and its result together 'similar' to the go and go_test files.

ezekiel commented:
I like it, that's actually how it was organized in early stages, but then the original tests were processing the `golden` files as if they were certificates.

I'll push changes with an update to the original tests to load `filepath.Glob("testdata/*.pem")` instead of the whole directory, and then the expected output files could be in the same directory.

We'll see how it looks.

cpu commented on certlint_test.go:
```diff
@@ -62,12 +90,117 @@ func TestTestData(t *testing.T) {
+				t.Fatal(err)
+			}
+
+			//if *update {
```
You have some commented cruft here that could be deleted.

ezekiel replied:
Ahh yes, I had that there to remind myself to implement something for programatically updating the `golden` files. I'll switch that to a `TODO` that matches the format seen in the other tests.

cpu commented on certlint_test.go:
```diff
@@ -46,12 +50,36 @@ accZ5e1iOtNIBTVUEaQ0YxNwhZqnQO4rgBFCnEpNH6TB62hXUD6u/oP/WB48Wmek
 KPgaoWHMo5sOFQnw5A==
 -----END CERTIFICATE-----`
 
+func TestMain(m *testing.M) {
```
Is there a reason you're invoking `exec.Command` to build the test data? I might be overlooking something but `go test ./...` should build as required no?

Same Q: re the `go clean` invocation. I'm not sure I understand why its needed. Maybe a comment would help explain?

ezekiel replied:
Unfortunately, no. It doesn't seem to automatically build a CLI binary for tests without this trickery in TestMain.

However, I am extremely new to writing tests for go (CLI tool or otherwise), so I would be elated if I were wrong and could delete the whole TestMain trickery. If it were actually building a binary, would it land in `~/.cache/go-build/`, or somewhere else? If it landed in the cache, how would I be able to predict the name in order to execute the CLI utility for testing?

Perhaps a point to be made here is that test code would be much more useful as further unit tests rather than CLI utility tests (essentially integration tests). Or, CLI utility tests would be much better as travis tests?

cpu replied:
Ahhh! I understand better now. I hadn't really clued in that this is more of an integration test and needs to run the project binaries.

I think one way to approach this case might be to have the `.travis.yml` `before_script` stage (or equiv) run `go install` on the project packages. The `certlint` binary should end up in `$GOPATH/bin` and I _think_ that's on the travis `$PATH` for Go projects by default. You would be able to remove the code you have to build the `certlint` binary and `certlint_test.go` would be just test code and nothing related to build/install.

I might be sending you on a wild goose chase based on vague theories :-) I'll try to actually test this when it's not Friday night. Def. don't consider my input here a blocker.

vanbroup replied:
Not use if adding a build to the `before_script` in `.travis.yml` would really help here, this would prevent you from running the test locally before committing. Unless you do the build manually.

cpu replied:
I did give this a short test in https://github.com/cpu/certlint/commit/beeb7241cde419179c5678e1be6a3cfd3d67922b and using a `go install` worked OK and seemed cleaner than the cross-concern of building/installing from within the test code.

RE: local testing, it seems reasonable to assume that someone who wants to run the integration tests should have the tool installed (e.g. has run `go install ./...` in the repo). You are correct that if the integration test didn't also build & install the binaries you wouldn't be able to run it without doing so yourself.

ezekiel commented:
Re-organized the `golden` files in the original commit (and rebased on updated master).

Also added a new commit that adds a single test to the CLIReturn test table. Pleased to see how simple it is to add another test for CLI utility return value.

vanbroup commented on certlint_test.go:
```diff
@@ -46,37 +51,166 @@ accZ5e1iOtNIBTVUEaQ0YxNwhZqnQO4rgBFCnEpNH6TB62hXUD6u/oP/WB48Wmek
+
+	retcode := m.Run()
+
+	clean := exec.Command("go", "clean", "-x")
```
Do you think `go clean` is required here? Normally `go` would take care of this automatically not?

ezekiel replied:
True that relying on Travis would mean that one would have to perform "go install" or at the very least "go build" before being able to run local CLI tests.

Unfortunately, when using TestMain to set things up for the test, one must also clean up. Without `go clean`, the binary is left hanging around in the base directory.

vanbroup commented on certlint_test.go:
```diff
@@ -46,37 +51,166 @@ accZ5e1iOtNIBTVUEaQ0YxNwhZqnQO4rgBFCnEpNH6TB62hXUD6u/oP/WB48Wmek
+	}
+	testdata := path.Join(dir, "testdata")
+	files, _ := filepath.Glob("./testdata/*.pem")
+	for _, f := range files {
```
Is it relevant to run the test against all test data again? The test data is a resource that should be growing to cover much more cases to make sure we detect and will always detect them, running these test twice looks to be a waste of resources and time.

Should we limit the test to a few relevant checks to just check if the CLI does what it should be doing?

ezekiel replied:
Yep, that approach sounds fine to me.

vanbroup commented on certlint_test.go:
```diff
@@ -46,37 +51,166 @@ accZ5e1iOtNIBTVUEaQ0YxNwhZqnQO4rgBFCnEpNH6TB62hXUD6u/oP/WB48Wmek
+				t.Fatal(err)
+			}
+
+			// TODO: allow for programatic refresh of golden files as a test argument
```
It would be great if you could add this before we merge this PR.

ezekiel replied:
Okay, I'll get that sorted out.

ezekiel replied:
I've added this in further commits.

ezekiel commented:
Just commenting here to say that I have not dropped this. I should have the time to finish the required updates in the next week or so.

ezekiel commented:
Hopefully this update addresses all outstanding concerns regarding the _local_ integration tests for the CLI binary.

vanbroup commented:
Thanks @ezekiel
